### PR TITLE
Dynamic plugins

### PR DESCRIPTION
**Pull request by mpangburn**
Enables changing the pump and CGM managers provided within Loop using a text file, PLUGINS, which contains a list of frameworks that should be linked.

The strategy is as follows:

- A new build target, Plugins, contains a script that writes to Loop+Plugins.xcconfig with the additional linker flags for the frameworks listed in PLUGINS prior to the linking phase. The frameworks are expected to exist in the Carthage/Build directory.
- The Info.plist file of each plugin framework describes the managers it provides using the keys "PumpManager" and/or "CGMManager". The value for each key can be either an individual string or an array of strings, where each string is the name of a class implementing the protocol described by that key.
- At runtime, Loop traverses the list of linked bundles for these class names and grabs their associated types using the Obj-C runtime (NSClassFromString).

As a consequence of this strategy, Loop should never need to `import ManagerProvidingFramework` explicitly.

### CODE REVIEW

**Comment by mpangburn** [comment]
The following items still require consideration:
- The LoopStatusExtension target generates some build-time warnings about unsafely linking against the UI frameworks. Still need to track down why that target is being linked against plugins.
- How significant is the imposed change on cross-framework development? Because Xcode no longer knows that Loop is linked against plugin frameworks, it will no longer automatically detect changes in those frameworks and rebuild when Loop builds.

File: Loop/Managers/CGMManager.swift [hunk 1]
```diff
@@ -6,18 +6,9 @@
...
 
 
-let allCGMManagers: [CGMManager.Type] = [
-    G6CGMManager.self,
-    G5CGMManager.self,
-    G4CGMManager.self,
-    ShareClientManager.self,
-]
-
+let allCGMManagers = Bundle.linkedCGMManagers.sorted(by: { $0.localizedTitle < $1.localizedTitle })
```
Comment (darinkrauss) [inline Loop/Managers/CGMManager.swift:11]: If we sort by localized title, will that give us the order we want? I assume we'd want the more recent CGMs (eg. G6) before the older ones.

File: Loop+Plugins.xcconfig [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#include "/Users/mpangburn/Developer/loop-priv/Loop.xcconfig"
```
Comment (darinkrauss) [inline Loop+Plugins.xcconfig:1]: Should this file be committed? It has your local machine path for one.

File: Common/Models/PumpManager.swift [hunk 3]
```diff
@@ -7,12 +7,9 @@
 
 import Foundation
 import LoopKit
-import MinimedKit
 
 
-let allPumpManagers: [PumpManager.Type] = [
-    MinimedPumpManager.self
-]
+let allPumpManagers = Bundle.linkedPumpManagers.sorted(by: { $0.localizedTitle < $1.localizedTitle })
```
Comment (darinkrauss) [inline Common/Models/PumpManager.swift:12]: If we sort by localized title, will that give us the order we want?

File: Common/Extensions/Bundle+LoopKit.swift [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import Foundation
+import LoopKit
+import LoopKitUI
+
+
+extension Bundle {
+    static let linkedPumpManagers = linkedFrameworkClasses(fromInfoDictionaryKey: "PumpManager").compactMap { $0 as? PumpManager.Type }
+    static let linkedCGMManagers = linkedFrameworkClasses(fromInfoDictionaryKey: "CGMManager").compactMap { $0 as? CGMManager.Type }
+
+    private static func linkedFrameworkClasses(fromInfoDictionaryKey plistKey: String) -> [AnyClass] {
```
Comment (darinkrauss) [inline Common/Extensions/Bundle+LoopKit.swift:18]: Clever. I like this idea.

Comment (darinkrauss) [reply Common/Extensions/Bundle+LoopKit.swift:18]: I might make this function public and move `linkedPumpManagers` and `linkedCGMManagers` into locations that are specific to PumpManagers and CGMManagers.

When I see an extension I ask myself "is this generally useful to other users of the thing being extended?" It seems like `linkedFrameworkClasses` is a general Bundle-related function that could be used in all sorts of Bundle situations. However, for most users of the general Bundle class, the `linkedPumpManagers` and `linkedCGMManagers` make no sense. However, `linkedPumpManagers` and `linkedCGMManagers` would make total sense in `DeviceDataManager` or similar location.

Or am I missing something? Thoughts?

**Review by darinkrauss — COMMENTED** [review]
I'm going to wait to approve this until after we figure out if we are going to go with the new app-level workspace and project.

**Comment by mpangburn** [comment]
Closing; this isn’t the right approach.

**Pull request closed without merging**